### ISSUE TITLE
RFC bib model

### ISSUE THREAD

**Issue by andrew2net** [relaton/relaton-ietf#1]
@ronaldtse do we have RNG grammas for RFC bibliographic item?

**Comment by ronaldtse** [comment]
@opoudjis could you help extend the current one for IETF documents? I suppose we should model it according to the attributes given in `<reference>` of RFC XML v3.

**Comment by opoudjis** [comment]
So, RFC XML already has a serialisation, as captured in https://tools.ietf.org/html/rfc7991 for the `reference` element. I take it this request is to express RFC in the isobib model. I *hope* this is for the purposes of citing RFC etc in ISO, and not in RFC, since the roundtrip is kinda silly.

OK, I'll get to mapping.

**Comment by ronaldtse** [comment]
Certainly it is for expressing RFC in the `<bibitem>` model: it should be used in any metanorma documents, including ISO, CSD, CSAND, etc.

**Comment by andrew2net** [comment]
@ronaldtse @opoudjis could we use XML from https://www.rfc-editor.org/refs/bibxml/reference.RFC.8341.xml for example? So we don't have to make model class and serialization, just get the XML document and include it to asciidoctor-gb output.

**Comment by ronaldtse** [comment]
No, the point is we need to translate from the IETF Bib XML format to the Relaton data model format. Asciidoctor-gb will not (and should not) understand the IETF format.

**Comment by opoudjis** [comment]
There may be refinements down the road, but this is an attempt to map all of the RFC XML to be found in the RFC references into IsoXML.

````
NOTE: This references the draft grammar at https://github.com/riboseinc/iso690xml/blob/master/grammars/biblio.rnc

reference@lang => bibitem/language
reference@base => IGNORE
reference@anchor => bibitem@id
reference@target => bibitem/source
reference@quoteTitle => IGNORE
reference/front/title => bibitem/btitle
reference/front/title@abbrev => IGNORE
reference/front/title@ascii => IGNORE
reference/front/seriesInfo@asciiName: see reference/front/seriesInfo@name. General rule for ascii* attributes: ignore in favour of UTF-8 originals
reference/front/seriesInfo@asciiValue: see reference/front/seriesInfo@value
reference/front/seriesInfo@name => bibitem/series/title
reference/front/seriesInfo@status => bibitem@type
reference/front/seriesInfo@stream => bibitem/contributor[role@type = "author"]/organization/name
reference/front/seriesInfo@value => bibitem/series/number
reference/front/author@asciiFullname: see reference/front/author@fullname
reference/front/author@asciiInitials: see reference/front/author@initials
reference/front/author@asciiSurname: see reference/front/author@surname
reference/front/author@fullname => bibitem/contributor[role@type = "author"]/person/name/completeName
reference/front/author@initials => bibitem/contributor[role@type = "author"]/person/name/initial
reference/front/author@role => bibitem/contributor/role@type : if this is populated, bibitem/contributor[role@type = "author"] becomes bibitem/contributor[role@type = "editor"]
reference/front/author@surname => bibitem/contributor[role@type = "author"]/person/name/surname
reference/front/author/organization => bibitem/contributor[role@type = "author"]/person/affiliation/organization/name : default is IETF
reference/front/author/organization@abbrev => bibitem/contributor[role@type = "author"]/person/affiliation/organization/abbreviation
reference/front/author/address/postal/postalLine => bibitem/contributor[role@type = "author"]/person/contact/street : interim solution, we haven't modelled this
reference/front/author/address/postal/postalLine@ascii: see reference/front/author/address/postal/postalLine
reference/front/author/address/postal/city => bibitem/contributor[role@type = "author"]/person/contact/city
reference/front/author/address/postal/city@ascii: see reference/front/author/address/postal/city
reference/front/author/address/postal/code => bibitem/contributor[role@type = "author"]/person/contact/postcode
reference/front/author/address/postal/code@ascii: see reference/front/author/address/postal/code
reference/front/author/address/postal/country => bibitem/contributor[role@type = "author"]/person/contact/country
reference/front/author/address/postal/country@ascii: see reference/front/author/address/postal/country
reference/front/author/address/postal/region => bibitem/contributor[role@type = "author"]/person/contact/state
reference/front/author/address/postal/region@ascii: see reference/front/author/address/postal/region
reference/front/author/address/postal/street => bibitem/contributor[role@type = "author"]/person/contact/street
reference/front/author/address/postal/street@ascii: see reference/front/author/address/postal/street
reference/front/author/address/phone => bibitem/contributor[role@type = "author"]/person/contact/phone
reference/front/author/address/facsimile : NOT MODELLED YET
reference/front/author/address/email => bibitem/contributor[role@type = "author"]/person/contact/email
reference/front/author/address/email@ascii: see reference/front/author/address/email
reference/front/author/address/uri => bibitem/contributor[role@type = "author"]/person/contact/uri
reference/front/date@day => bibitem/date[@type = "published"]/on (compose)
reference/front/date@month => bibitem/date[@type = "published"]/on (compose)
reference/front/date@year => bibitem/date[@type = "published"]/on (compose)
reference/front/area => bibitem/classification[@type = "area"]
reference/front/workgroup  => bibitem/contributor[role@type = "author"]/person/affiliation/organization/subdivision
reference/front/keyword => bibitem/classification[@type = "area"]
reference/front/note => bibitem/note (header: NOTE)
reference/front/boilerplate => IGNORE
reference/annotation => bibitem/note (header: ANNOTATION)
reference/format => IGNORE
reference/refcontent => bibitem/note (header: REFCONTENT)
reference/seriesInfo: see reference/front/seriesInfo

Text formatting

reference/front/abstract@anchor => IGNORE
reference/front/abstract/dl => dt
reference/front/abstract/dl/dt => dt
reference/front/abstract/dl/dd => dd
reference/front/abstract/dl@anchor => IGNORE
reference/front/abstract/dl@hanging => IGNORE
reference/front/abstract/dl@spacing => IGNORE
reference/front/abstract/ol => ol
reference/front/abstract/ol@anchor => IGNORE
reference/front/abstract/ol@group => IGNORE
reference/front/abstract/ol@spacing => IGNORE
reference/front/abstract/ol@start => IGNORE
reference/front/abstract/ol@type => ol@type (with a mapping)
reference/front/abstract/ol/li => li
reference/front/abstract/ul => ul
reference/front/abstract/ul@anchor => IGNORE
reference/front/abstract/ul@empty => IGNORE
reference/front/abstract/ul@spacing => IGNORE
reference/front/abstract/ul/li => li
reference/front/abstract/t => p
reference/front/abstract/t@anchor => IGNORE
reference/front/abstract/t@hangText => IGNORE
reference/front/abstract/t@keepWithNext => IGNORE
reference/front/abstract/t@keepWithPrevious => IGNORE
reference/front/abstract/.../artwork => image
reference/front/abstract/.../artwork@align => IGNORE
reference/front/abstract/.../artwork@alt => IGNORE
reference/front/abstract/.../artwork@anchor => IGNORE
reference/front/abstract/.../artwork@height => image@height
reference/front/abstract/.../artwork@name => IGNORE
reference/front/abstract/.../artwork@src => image@src
reference/front/abstract/.../artwork@type => IGNORE
reference/front/abstract/.../artwork@width => image@width
reference/front/abstract/.../artwork@xml:space => IGNORE
reference/front/abstract/.../figure => figure
reference/front/abstract/.../figure@align => IGNORE
reference/front/abstract/.../figure@alt => IGNORE
reference/front/abstract/.../figure@anchor => IGNORE
reference/front/abstract/.../figure@height => IGNORE
reference/front/abstract/.../figure@src => IGNORE
reference/front/abstract/.../figure@suppress-title => IGNORE
reference/front/abstract/.../figure@title => figure/name
reference/front/abstract/.../figure@width => IGNORE
reference/front/abstract/.../figure/name => figure/name
reference/front/abstract/.../figure/preamble => .../p
reference/front/abstract/.../figure/postamble => .../p
reference/front/abstract/.../sourcecode => sourcecode
reference/front/abstract/.../sourcecode@anchor => IGNORE
reference/front/abstract/.../sourcecode@name => IGNORE
reference/front/abstract/.../sourcecode@src => IGNORE
reference/front/abstract/.../sourcecode@type => IGNORE
reference/front/abstract/.../bcp14A => text
reference/front/abstract/.../cref => review
reference/front/abstract/.../cref@anchor => review/id
reference/front/abstract/.../cref@display => IGNORE
reference/front/abstract/.../cref@source => review/reviewer
reference/front/abstract/.../em => em
reference/front/abstract/.../eref => text
reference/front/abstract/.../eref@target => link
reference/front/abstract/.../iref => IGNORE
reference/front/abstract/.../list => ul/li
reference/front/abstract/.../relref => eref
reference/front/abstract/.../relref@displayFormat => IGNORE
reference/front/abstract/.../relref@relative => IGNORE
reference/front/abstract/.../relref@section => eref/locality[@type = "section"]
reference/front/abstract/.../relref@target => xref@target
reference/front/abstract/.../spanx => text (there's a lot more formatting, but this is legacy v2 element)
reference/front/abstract/.../strong => strong
reference/front/abstract/.../sub => sub
reference/front/abstract/.../sup => sup
reference/front/abstract/.../tt => tt
reference/front/abstract/.../vspace => br
reference/front/abstract/.../xref => xref
reference/front/abstract/.../xref@format => IGNORE
reference/front/abstract/.../xref@pageno => IGNORE
reference/front/abstract/.../xref@target => xref@target
````

**Comment by opoudjis** [comment]
@andrew2net, I'm trying to introduce release notes into all Metanorma gems, now that they are starting to be used in production. If you bump one of your gems related to Metanorma, please create a release with notes that summarise all changes, and link to related tickets. This issue is now linked to https://github.com/riboseinc/iso-bib-item/releases/tag/v0.1.7

**Comment by ronaldtse** [comment]
@opoudjis are we going to do an XSLT or a Ruby implementation for RelatonXML to RFC BibXML?

**Comment by andrew2net** [comment]
@opoudjis it could be several `reference/front/seriesInfo`. So, how we shoud assign multiple `series@info` to one `bibitem@type?
```
reference/front/seriesInfo@status => bibitem@type
```

**Comment by opoudjis** [comment]
RFC has an idiosyncratic way of dealing with seriesInfo, it encodes two different pieces of information. I will summarise later today.

**Comment by opoudjis** [comment]
We have two versions of RFC XML, unfortunately: v2 https://tools.ietf.org/html/rfc7749 and v3 https://tools.ietf.org/html/rfc7991#section-2.47. The v2 tags are included as v3 as deprecated elements, and I have included both in the mappings above.

The RFC XML currently on https://xml2rfc.tools.ietf.org/public/rfc/bibxml is RFC XML v2. In fact, support for v3 is still quite poor, and v3 may yet be radically altered. Which is a pity, because overall v3 is better—except for its handling of series.

In v2, seriesInfo is a simple tag, that indicates only the name and value of the series as attributes. In https://xml2rfc.tools.ietf.org/public/rfc/bibxml/reference.RFC.0001.xml, one series is the RFC series, which is the one we are interested in: 

`<seriesInfo name="RFC" value="1"/>` => `<bibitem><series><title>RFC</title><number>1</number></bibitem>`

The second series given in that URL is the DOI of the document, and that is actually not a series, it's a document identifier:  `<seriesInfo name="DOI" value="10.17487/RFC0001"/>` =>  `<bibitem><docidentifier type="DOI"> 10.17487/RFC0001 </docidentifier></bibitem>`.

So for the documents you're going to be getting from the https://xml2rfc.tools.ietf.org sites, the best thing to do is to have a different mapping for each seriesInfo@name type; some will be better treated as series, some as doc identifiers. I'll go further though: properly we should be providing only one series for each document, and any other seriesInfo instances should be treated as document identifiers. If a document is somehow published by two organisations, each organisation can assign it to its own series, but that is unusual (though in the standards world, probably not unusual enough). 

All of these elements:

````
reference/front/seriesInfo@asciiName: see reference/front/seriesInfo@name. General rule for ascii* attributes: ignore in favour of UTF-8 originals
reference/front/seriesInfo@asciiValue: see reference/front/seriesInfo@value
reference/front/seriesInfo@status => bibitem@type
reference/front/seriesInfo@stream => bibitem/contributor[role@type = "author"]/organization/name
reference/front/seriesInfo@value => bibitem/series/number
````

are from RFC XML v3. v3 does something very annoying: 

>  If a <front> element contains a <seriesInfo> element with a name
      of "Internet-Draft", it can also have at most one additional
      <seriesInfo> element with a "status" attribute whose value is of
      "standard", "full-standard", "bcp", "fyi", "informational",
      "experimental", or "historic" to indicate the intended status of
      this Internet-Draft, if it were to be later published as an RFC.
      If such an additional <seriesInfo> element has one of those
      statuses, the name needs to be "".

>  If a <front> element contains a <seriesInfo> element with a name
      of "RFC", it can also have at most one additional <seriesInfo>
      element with a "status" attribute whose value is of
      "full-standard", "bcp", or "fyi" to indicate the current status of
      this RFC.  If such an additional <seriesInfo> element has one of
      those statuses, the "value" attribute for that name needs to be
      the number within that series.  That <front> element might also
      contain an additional <seriesInfo> element with the status of
      "info", "exp", or "historic" and a name of "" to indicate the
      status of the RFC.

In other words, even if there are multiple seriesInfo tags per document, RFC expects only one of those tags to contain a status attribute. The status attribute does not really belong in RFC XML, and I wouldn't be that concerned if you omit it from the mapping to Relaton XML completely; but the issue of multiple seriesInfo@status instances is not intended to apply. And in any case, this is RFC XML v3, which I doubt is going to be available for online bibliographies for some time.

**Comment by opoudjis** [comment]
Found an instance where an item is actually cited as belonging to two different series by two different organisations:

https://xml2rfc.tools.ietf.org/public/rfc/bibxml2/_reference.ITU.X530.2001.xml

````
<seriesInfo name="ITU-T" value="Recommendation X.530"/>
<seriesInfo name="ISO" value="Standard 9594-10"/>
````

That is.... annoying. I think we really do have to model this as two distinct `series` elements in the bibitem. But the issue of ambiguity of seriesInfo@status does not arise in these: seriesInfo@status is specific to RFC and Internet-Draft series, published by the IETF.
